Создан слой DAO

xma4602 opened the pull request.

xma4602 committed:
Первая итерация системы.

xma4602 committed:
Ошибки перенесены в другой пакет

xma4602 committed:
Сущности переписаны и помещены в отдельный пакет

xma4602 committed:
Добавлены репозитории сущностей.

xma4602 committed:
Добавлены сервисы сущностей.
Сделан ленивый генератор номеров.

xma4602 committed:
Изменены зависимости в pom.xml
Запросы инициализации таблиц перенесены в один запрос.

xma4602 committed:
Отредактированы сущности. Добавлен двусторонний меппинг один ко многим.

xma4602 committed:
Отредактированы сервисы из-за появления меппинга. Теперь сервисы клиента и счета знают только о своих сущностях и репозиториях.

xma4602 committed:
Переделаны методы сервисов. Теперь они работают только с номерами сущностей.

xma4602 committed:
Испралена кодировка сообщений в ошибках. + небольшие изменения

xma4602 committed:
Изменен формат работы сервисов клиента и счета.

xma4602 committed:
Добавлен сервис финансовых операций.
Добавлен сервис пользовательских операций.

TumasovAM commented on src/main/java/com/bank/entities/Account.java:
```diff
@@ -0,0 +1,69 @@
+
+    @Id
+    @GeneratedValue
+    protected UUID id;
```
Никакой атрибут класс не должен быть получен напрямую никогда

TumasovAM commented on src/main/java/com/bank/errors/InsufficientFundsException.java:
```diff
@@ -0,0 +1,18 @@
+
+    @Override
+    public String getMessage() {
+        return "На счете №%d недостаточно средств, чтобы снять сумму %d".formatted(accountNumber, sum);
```
Зачем кириллица? Кто её будет видеть?

TumasovAM commented on src/main/java/com/bank/servises/AccountService.java:
```diff
@@ -0,0 +1,59 @@
+
+    public Account getAccount(long accountNumber) throws NoSuchAccountException {
+        var result = accountRepository.findAccountByNumber(accountNumber);
+        if (result.isEmpty()) throw new NoSuchAccountException(accountNumber);
```
Не по конвенции

if(true){
...
}

TumasovAM commented on src/main/java/com/bank/servises/AccountService.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public Account getAccount(long accountNumber) throws NoSuchAccountException {
+        var result = accountRepository.findAccountByNumber(accountNumber);
```
var лучше не использовать. Не читаемый код, особенно при ревью.

TumasovAM commented on src/main/java/com/bank/servises/AccountService.java:
```diff
@@ -0,0 +1,59 @@
+        accountRepository.save(account);
+        return account;
+    }
+    public void addAccount(Account account) {
```
addAccount - добавить счет. Метод же создает новый счет. Это разные вещи

TumasovAM commented on src/main/java/com/bank/servises/AccountService.java:
```diff
@@ -0,0 +1,59 @@
+        return accountRepository.findAccountsByClientNumber(clientNumber);
+    }
+
+    public Account addAccount(Client client) {
```
Добавление счета к клиенту не зона ответственности этого сервиса

xma4602 replied:
Почему не его ответственность? AccountService отвечает за создание и удаление записей банковских счетов - в том числе создание нового счета для клиента. Разве это смешение функционала?

TumasovAM commented on src/main/java/com/bank/servises/ClientService.java:
```diff
@@ -0,0 +1,54 @@
+            clientRepository.delete(client);
+            return true;
+        } catch (NoSuchClientException e) {
+            return false;
```
Нельзя скрывать Exception. Обязательно писать в логи

TumasovAM commented on src/main/java/com/bank/servises/FinanceService.java:
```diff
@@ -0,0 +1,58 @@
+        accountService.addAccount(account);
+    }
+
+    public void takeFrom(long accountNumber, long sum) throws NoSuchAccountException, InsufficientFundsException {
```
Метод вне транзакции. Пока метод высчитывает можно или нельзя, с балансом могут происходить тысячи операций. В любую сторону

TumasovAM commented on src/main/java/com/bank/servises/FinanceService.java:
```diff
@@ -0,0 +1,58 @@
+        accountService.addAccount(account);
+    }
+
+    public void transfer(long senderAccountNumber, long recipientAccountNumber, long sum) throws NoSuchAccountException, InsufficientFundsException {
```
Метод вне транзакции

xma4602 committed:
Сообщения ошиборк переписаны на английском.

xma4602 committed:
Поля сущностей сделаны приватными. Зависимости с другими сущностями сделаны LAZY.

xma4602 committed:
В сервисам var заменены явными типами. К if добавлены кобки. Переименовы некоторые методы.

xma4602 committed:
Зависимые изменения из прошлого коммита: переименования методов.

xma4602 committed:
Система переведена на использование UUID

xma4602 committed:
Небольшие изменения репозиториев

xma4602 committed:
В AccountService переработаны методы создания, удаления и изменения.
FinanceService переработаны методы пополения, снятия и перевода средсв счета.
UserService изменен в соотвествии с корректировками предыдущих сервисов.

TumasovAM commented on src/main/java/com/bank/entities/Client.java:
```diff
@@ -0,0 +1,72 @@
+
+    @Id
+    @GeneratedValue
+    private UUID id;
```
Хорошей практикой считается если сущность наследует базовый класс. В него принято выносить как минимум id.

TumasovAM commented on src/main/java/com/bank/repositories/AccountRepository.java:
```diff
@@ -0,0 +1,22 @@
+
+    void deleteAccountById(UUID id);
+
+    void updateById(UUID id, long amount);
```
1) void допускается только для delete метода
2) Для того чтобы обновить сущность принято передавать всю сущность. Даже если меняется только один атрибут. Сейчас ты решил менять только amount. А завтра захочешь менять amount и name сразу. Делать третий метод update?

TumasovAM commented on src/main/java/com/bank/repositories/ClientRepository.java:
```diff
@@ -0,0 +1,18 @@
+
+    void deleteBy(UUID id);
+
+    long updateById(UUID id, String name);
```
Любой метод кроме delete должен отдавать саму сущность.
